aten/src/ATen/native/quantized/cpu/q_avgpool.cpp

```diff
@@ -12,212 +13,224 @@ namespace at {
+          (ptr_output+i)->val_ = 0;
+
+        for(yy = 0; yy < outputHeight; yy++)
+        {
+          for(xx = 0; xx < outputWidth; xx++)
+          {
+            /* Compute the mean of the input image... */
+            int64_t hstart = yy * dH - padH;
+            int64_t wstart = xx * dW - padW;
+            int64_t hend = std::min(hstart + kH, inputHeight + padH);
+            int64_t wend = std::min(wstart + kW, inputWidth + padW);
+            int pool_size = (hend - hstart) * (wend - wstart);
+            hstart = std::max(hstart, (int64_t) 0);
+            wstart = std::max(wstart, (int64_t) 0);
+            hend = std::min(hend, inputHeight);
+            wend = std::min(wend, inputWidth);
+
+            int64_t sum = 0;
+
+            int divide_factor;
+            if (divisor_override.has_value()) {
+              divide_factor = divisor_override.value();
+            } else {
+              if(count_include_pad) {
+                divide_factor = pool_size;
+              } else {
+                divide_factor = (hend - hstart) * (wend - wstart);
+              }
+            }
 
-inline int end_index(int a, int b, int c) {
-  return (int)std::ceil((float)((a + 1) * c) / b);
-}
+            int64_t kx, ky;
 
-template <typename scalar_t, typename underlying_t>
-static void adaptive_avg_pool2d_single_out_frame(
-    scalar_t* input_p,
-    scalar_t* output_p,
-    int64_t sizeD,
-    int64_t isizeH,
-    int64_t isizeW,
-    int64_t osizeH,
-    int64_t osizeW,
-    int64_t istrideD,
-    int64_t istrideH,
-    int64_t istrideW) {
-  at::parallel_for(0, sizeD, 0, [&](int64_t start, int64_t end) {
-    for (auto d = start; d < end; d++) {
-      /* loop over output */
-      int64_t oh, ow;
-      for (oh = 0; oh < osizeH; oh++) {
-        int istartH = start_index(oh, osizeH, isizeH);
-        int iendH = end_index(oh, osizeH, isizeH);
-        int kH = iendH - istartH;
-        float kHr = 1.0 / kH;
-
-        for (ow = 0; ow < osizeW; ow++) {
-          int istartW = start_index(ow, osizeW, isizeW);
-          int iendW = end_index(ow, osizeW, isizeW);
-          int kW = iendW - istartW;
-          float kHWr = kHr / kW;
-
-          /* local pointers */
-          scalar_t* ip =
-              input_p + d * istrideD + istartH * istrideH + istartW * istrideW;
-          scalar_t* op = output_p + d * osizeH * osizeW + oh * osizeW + ow;
-
-          /* compute local average: */
-          int64_t sum = 0;
-          int ih, iw;
-          for (ih = 0; ih < kH; ih++) {
-            for (iw = 0; iw < kW; iw++) {
-              int64_t val = (ip + ih * istrideH + iw * istrideW)->val_;
-              sum += val;
+            for(ky = hstart; ky < hend; ky++)
+            {
+              for(kx = wstart; kx < wend; kx++)
+                sum += (ptr_input + ky*inputWidth + kx)->val_;
             }
+            /* Update output */
+            ptr_output->val_ = static_cast<underlying_t>(std::nearbyint(sum/divide_factor));
```

Comment:
`sum/divide_factor` is done in integer arithmetic. I assume that's not intended?

Reply:
That's a good point, The divide should be done in float arithmetic and then rounded to the nearest int, currently both sum and divide_factor are ints.